types/xdg_shell/wlr_xdg_surface.c

```diff
@@ -507,6 +508,11 @@ void reset_xdg_surface(struct wlr_xdg_surface *xdg_surface) {
 	}
 
 	xdg_surface->role = WLR_XDG_SURFACE_ROLE_NONE;
+
+	wl_list_remove(&xdg_surface->surface_destroy.link);
+	wl_list_remove(&xdg_surface->surface_commit.link);
```

Comment:
Doesn't this break re-mapping a surface after it has been unmapped?

Reply:
I also thought this but after grepping in the source code, you can see `reset_xdg_surface` is called in three cases:

- destruction of xdg_toplevel
- destruction of xdg_popup
- destruction of xdg_surface

So this is not a problem

Reply:
The client is allowed to destroy the xdg_toplevel object without destroying the xdg_surface, and then re-create the xdg_toplevel.

The `surface_commit` listener is set up in `create_xdg_surface`, I'd rather not remove it in `reset_xdg_surface`.